build: add authors.yml

bnb opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

This PR adds a GitHub Action, `authors.yml`, that be triggered either automatically on the first day of the month at 00:00:00 (likely GMT but maybe EST?) or manually (via GitHub Actions UI).

Specifically, this action will run `tools/update-authors.js`. If there are changes, it will either create a new Pull Request or update an `open` Pull Request that the Action already created. Some additional things to note:

- This Action uses @gr2m's [create-or-update-pull-request-action](https://github.com/gr2m/create-or-update-pull-request-action).
- This bot will create a branch `actions/authors-update` where it'll do the work. I specifically prefer this namespace layout for clarity when scrolling through branches, but if a different namespacing scheme is desired I'm happy to tweak the config. 
- I've set it to label the `AUTHORS` file update PRs with the `meta` label. If there's a more appropriate label/set of labels, please feel free to let me know and I'll tweak the config.

To see an example of what the PRs will look like, check out [the PR on my fork](https://github.com/bnb/node/pull/3).

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

bnb commented:
cc @jasnell since this was a request from you ❤️

silverwind commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,23 @@
+on:
+  schedule:
+    # 00:00:00 on the first day of every month
+    # https://crontab.guru/#0_0_1_1-12_*
+    - cron: "0 0 * 1-12 *"
```
```suggestion
    # https://crontab.guru/#0_0_1_*_*
    - cron: "0 0 * * *"
```

bnb replied:
Description for the current:

> At 00:00 on day-of-month 1 in every month from January through December.

Description for the change:

> At 00:00 on day-of-month 1 in every month from January through December.

They do the same thing, but `*` is a bit cleaner.

bnb replied:
Ah, I see the issue - the actual code was wrong but the link was right 😅

silverwind commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,23 @@
+    # 00:00:00 on the first day of every month
+    # https://crontab.guru/#0_0_1_1-12_*
+    - cron: "0 0 * 1-12 *"
+  workflow_dispatch:
```
What's this for?

gr2m replied:
it allows maintainers to trigger the action manually. You'll get a button like this

<img width="975" alt="image" src="https://user-images.githubusercontent.com/39992/97344760-708d0700-1846-11eb-8ede-6b30566905b3.png">

(without the "Version of https://www.npmjs.com/package/@github/openapi" input). See https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#manual-events

GaryGSC replied:
It adds a manual trigger that looks like [this](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/).

silverwind replied:
Ah, that's certainly useful, thanks.

lpinca commented:
It seems this action only runs `./tools/update-authors.js`. It is also necessary to remove duplicates and update `.mailmap` but and that is a manual operation. I don't know if that can be automated with the current `update-authors.js` script.

bnb commented:
@lpinca that's totally fair. The nice thing about how create-or-update-pull-request works is that it'll open a PR that can have changes to `.mailmap` manually pushed. If this work goes uncompleted until the next time the Action runs, the Action will simply update the _existing_ PR so there's not a massive backlog of PRs doing the same work.

Would it be useful to note that `.mailmap` will need to be manually updated in the PR body text?

lpinca commented:
Yes I think it makes sense to specify that there might be duplicates to be remove manually.

bnb commented:
Converted to draft. Hit a snag with the branch stuff that I've logged in the upstream Action's repo - once it's fixed (either because I use it correctly or released upstream) I'll un-draft this PR 👍🏻 

here's that issue: https://github.com/gr2m/create-or-update-pull-request-action/issues/245

FWIW @lpinca I've gone ahead and updated the metadata file to include a message about that in the PR body. Once the issue is addressed we should be good to go.

lpinca commented:
> FWIW @lpinca I've gone ahead and updated the metadata file to include a message about that in the PR body. Once the issue is addressed we should be good to go.

Not sure if you forgot to push a commit but I can't see it.

bnb commented:
@lpinca I committed it to `master` in my fork but not the branch 😅

lpinca commented:
It seems suggestion are not allowed in draft PRs.

I would replace the following sentence

"Please note, prior to merging please ensure there are no duplicate entries. If there are, please commit updates to `.mailmap` directly to this PR."

with

"Please note that there might be duplicate entries. If there are, please remove them and add the duplicate email/s to `.mailmap` directly to this PR."

bnb commented:
@lpinca done 👍🏻

bnb commented:
It seems the issue has been fixed (fix: https://github.com/gr2m/create-or-update-pull-request-action/pull/256, run with fix: https://github.com/bnb/node/actions/runs/366316485, PR example: https://github.com/bnb/node/pull/5).

This can presumably proceed.

targos commented:
@bnb is it expected that the test PR in https://github.com/bnb/node/pull/5 removes all authors?

bnb commented:
@targos it should not, no. Including `fetch-depth: '0'` reduced the number of folks being nuked from the list from 3000 to 700 an added the correct number of folks, but for some reason it's dropping everyone up to _me_ in that list. I cannot fathom why, trying to figure it out now but largely seeming to fail at that.

Trott commented:
Maybe we can land this with the `schedule:` stuff removed so it's an action that can only be triggered manually. Then we can test it a bunch manually, debug/fix any problems, and once we feel good about it, add the schedule stuff back in. Does that work for you, @bnb?

bnb commented:
Sounds good to me @Trott - I can spend some time today cleaning it up since IIRC the branch stuff was the biggest problem.

bnb commented:
Updated, works as per https://github.com/bnb/node/pull/7. We can probably merge and the immediately add the cron tbh.

Mesteery commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,21 @@
+name: "authors update"
```
```suggestion
name: Update AUTHORS
```

bnb replied:
I disagree with this, since it's following a similar pattern to https://github.com/nodejs/node/actions/workflows/license-builder.yml which uses `<name> update`.

Mesteery replied:
Okay! But maybe the (useless) quotes could be removed?

Mesteery commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,21 @@
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: '0' # this is required to actually get all the authors
```
```suggestion
          fetch-depth: 0 # this is required to actually get all the authors
```

bnb replied:
I don't think this is a nit worth retesting the script's functionality for. It works as is, and I'd prefer not to rock the boat with changes.

Mesteery replied:
There is no need to retest, unless you have a doubt, in which case I think a simple YAML validator is sufficient. Otherwise, removing unnecessary quotes might unload the action and make it a bit more readable. Personally, I don't see any problem with that.

Mesteery replied:
There is no need to retest, unless you have a doubt, in which case I think a simple YAML validator is sufficient. Otherwise, removing unnecessary quotes could make the workflow a bit more readable. Personally, I don't see any problem with that.

Mesteery commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,21 @@
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: '0' # this is required to actually get all the authors
+      - run: "tools/update-authors.js" # run the AUTHORS tool
```
```suggestion
      - run: tools/update-authors.js # run the AUTHORS tool
```

bnb replied:
See previous comment (https://github.com/nodejs/node/pull/35831#discussion_r666405084)

Mesteery commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,21 @@
+        with:
+          fetch-depth: '0' # this is required to actually get all the authors
+      - run: "tools/update-authors.js" # run the AUTHORS tool
+      - uses: gr2m/create-or-update-pull-request-action@v1.x # create a PR or update the Action's existing PR
```
```suggestion
      - uses: gr2m/create-or-update-pull-request-action@v1 # create a PR or update the Action's existing PR
```

bnb replied:
See previous comment (https://github.com/nodejs/node/pull/35831#discussion_r666405084).

Mesteery commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,21 @@
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          title: "doc: update AUTHORS"
+          body: "If this PR exists, there's presumably new additions to the AUTHORS file. This is an automatically generated PR by the `authors.yml` GitHub Action, which runs `tools/update-authors.js` and submits a new PR or updates an existing PR.\n\nPlease note that there might be duplicate entries. If there are, please remove them and add the duplicate emails to .mailmap directly to this PR."
```
```suggestion
          body: If this PR exists, there's presumably new additions to the AUTHORS file. This is an automatically generated PR by the `authors.yml` GitHub Action, which runs `tools/update-authors.js` and submits a new PR or updates an existing PR.\n\nPlease note that there might be duplicate entries. If there are, please remove them and add the duplicate emails to .mailmap directly to this PR.
```

bnb replied:
See previous comment (https://github.com/nodejs/node/pull/35831#discussion_r666405084)

Trott commented:
Not in this PR, but maybe we need a YAML linter for our expanding suite of GitHub Actions workflows.

bnb commented:
> Not in this PR, but maybe we need a YAML linter for our expanding suite of GitHub Actions workflows.

agree. Happy to take that on (trying to get more involved in core <3).

bnb commented:
force pushed with the version change and the working changes to get the commit message passing (AFAIK? still seems mad at me)

Trott commented:
> force pushed with the version change and the working changes to get the commit message passing (AFAIK? still seems mad at me)

`ci(automation):` is not a subsytem label we use. Try `build:` instead.

Valid subsystems are (probably not-entirely-correctly) listed at https://github.com/nodejs/node/blob/e3ba96053e062703b24715d64a7c6627e391fce7/doc/guides/contributing/pull-requests.md#appendix-subsystems. More canonically, you can check https://github.com/nodejs/core-validate-commit/blob/7b97be31b9036613ac17780ced62a2b1f8171a17/lib/rules/subsystem.js for the list of subsystems that the linter will accept.

bnb commented:
@Trott should I just force push the changes with the correct subsystem or...?

Trott commented:
> @Trott should I just force push the changes with the correct subsystem or...?

Yes, please! Thanks!

bnb commented:
@Trott done :)

bnb commented:
It is still failing despite there only being one commit with _seemingly_ the correct subsystem prefix? 

![image](https://user-images.githubusercontent.com/502396/130283649-353376b0-9263-43ec-9faf-49066088cf6b.png)

Trott commented:
> It is still failing despite there only being one commit with _seemingly_ the correct subsystem prefix?

I tried re-running the GitHub Action and it still failed. Looks like it's failing on some old commits that are in the rebase that have an outdated subsystem name.  A rebase and force push would likely fix the issue. I'll try that and we'll see.

(waits a few minutes to see results)

Yup, that worked!

Trott commented on .github/workflows/authors.yml:
```diff
@@ -0,0 +1,21 @@
+          title: "doc: update AUTHORS"
+          body: "If this PR exists, there's presumably new additions to the AUTHORS file. This is an automatically generated PR by the `authors.yml` GitHub Action, which runs `tools/update-authors.js` and submits a new PR or updates an existing PR.\n\nPlease note that there might be duplicate entries. If there are, please remove them and add the duplicate emails to .mailmap directly to this PR."
+          branch: "actions/authors-update" # custom branch *just* for this Action.
+          commit-message: "doc: update AUTHORS"
```
`doc` is fine but I'd prefer `meta`. The AUTHORS list really isn't documentation, at least not in the sense that a user would care about. The subsystem, as far as I know, mostly just determines where the commit ends up in the list of commits in the release notes. And I'd prefer this to be lumped in with the other meta stuff and not the doc stuff. If you disagree at all, feel free to leave it as `doc`. I'm fine with that. But if it makes no difference to you either way, I'd go with `meta` instead.

bnb replied:
done :)

Trott replied:
@bnb Looks like you changed it in the title (line 17) but not the commit-message (line 20). Can you change it in the commit message, squash it all down to one commit, and then force push? Once that's done, I'll add the commit-queue label and we can hopefully get this finally landed.

Trott replied:
No big deal for this, but in the future, if you could use [`--fixup` commits](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt) or else squash everything, it would mean we could land this by simply applying a label. If we apply the label now, the three commits will all land, and that's not really what we want.

Trott commented:
Landed in 02c5120d2a58
